crypto/s2n_certificate.c

```diff
@@ -605,3 +611,113 @@ int s2n_get_cert_der(const struct s2n_cert *cert, const uint8_t **out_cert_der,
 
     return S2N_SUCCESS;
 }
+
+static int s2n_asn1_string_free(ASN1_STRING** data)
+{
+    if (*data != NULL) {
+        ASN1_STRING_free(*data);
+    }
+    return S2N_SUCCESS;
+}
+
+int s2n_get_utf8_string_from_extension_data(const uint8_t *extension_data, uint32_t extension_len, uint8_t **out_data, uint32_t *out_len)
+{
+    POSIX_ENSURE_REF(extension_data);
+    POSIX_ENSURE_GT(extension_len, 0);
+    POSIX_ENSURE_REF(out_data);
+    POSIX_ENSURE_REF(out_len);
+
+    DEFER_CLEANUP(ASN1_STRING *asn1_str = NULL, s2n_asn1_string_free);
+    const uint8_t *asn1_str_data = extension_data;
+    /* Note that d2i_ASN1_UTF8STRING increments *der_in to the byte following the parsed data.
+     * Ref: https://www.openssl.org/docs/man1.1.0/man3/d2i_ASN1_UTF8STRING.html.
+     */
+    asn1_str = d2i_ASN1_UTF8STRING(NULL, (const unsigned char **)(void *)&asn1_str_data, extension_len);
+    POSIX_ENSURE(asn1_str != NULL, S2N_ERR_INVALID_X509_EXTENSION_TYPE);
+    int type = ASN1_STRING_type(asn1_str);
+    POSIX_ENSURE(type == V_ASN1_UTF8STRING, S2N_ERR_INVALID_X509_EXTENSION_TYPE);
+    
+    *out_len = ASN1_STRING_length(asn1_str);
+    POSIX_ENSURE_GT(*out_len, 0);
+
+    /* ASN1_STRING_data() returns an internal pointer to the data. 
+     * Since this is an internal pointer it should not be freed or modified in any way.
+     * Ref: https://www.openssl.org/docs/man1.0.2/man3/ASN1_STRING_data.html.
+     */ 
+    unsigned char *internal_data = ASN1_STRING_data(asn1_str);
+    POSIX_ENSURE_REF(internal_data);
+    POSIX_CHECKED_MEMCPY(*out_data, internal_data, (*out_len));
+
+    return S2N_SUCCESS;
+}
+
+int s2n_get_x509_extension_oid_value(struct s2n_cert *cert, const uint8_t *oid_field_in, const uint32_t oid_field_in_len,
+                                      uint8_t **oid_value_out, uint32_t *oid_value_out_len, bool *critical)
+{
+    POSIX_ENSURE_REF(cert);
+    POSIX_ENSURE_REF(oid_field_in);
+    POSIX_ENSURE_GT(oid_field_in_len, 0);
+    POSIX_ENSURE_REF(oid_value_out);
+    POSIX_ENSURE_REF(oid_value_out_len);
+    POSIX_ENSURE_REF(critical);
+    POSIX_ENSURE_REF(cert->raw.data);
+
+    uint8_t *der_in = cert->raw.data;
+    /* Obtain the openssl X509 cert from the ASN1 DER certificate input. 
+     * Note that d2i_X509 increments *der_in to the byte following the parsed data.
+     * Ref: https://www.openssl.org/docs/man1.1.0/man3/d2i_X509.html.
+     */
+    DEFER_CLEANUP(X509 *x509_cert = d2i_X509(NULL, (const unsigned char **)(void *)&der_in, cert->raw.size),
```

Comment:
Why all of the casting here? I would think it would work without it

Reply:
This casting is required here as the input required for `d2i_TYPE` is `const unsigned char **` and `der_in` is of type `uint8_t *` . If we remove the `void *` here we get an error. Reference to other location in s2n which has this casting: 

https://github.com/aws/s2n-tls/blob/fe8df74ddafb712d3141056107db7228b19b6ef7/crypto/s2n_pkey.c#L189